Add support for problem-specific sources in N Euler

carmaza opened the pull request:
Some tests in Newtonian hydro require problem-specific sources. This PR adds support for that. As a simple example, this PR adds the sources generated by a uniform acceleration (e.g. constant gravity), which are useful when studying non-linear instabilities (e.g Rayleigh-Taylor inst.). Other sources to be added in the future include central object potentials and "stirring" functions that drive turbulence.

- [ ] Bugfix
- [x ] New feature

### Component:

- [x ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

carmaza committed:
Add UniformAcceleration source for NewtonianEuler

carmaza committed:
Add Tags for problem-specific sources for NewtonianEuler

carmaza committed:
Add ComputeSources for NewtonianEuler

  - Add support for problem-specific source terms
  - ComputeSources is templated in source term type

wthrowe commented on src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.cpp:
```diff
@@ -0,0 +1,51 @@
+
+template <size_t Dim>
+UniformAcceleration<Dim>::UniformAcceleration(
+    std::array<double, Dim> acceleration_field) noexcept
```
const

wthrowe commented on src/Evolution/Systems/NewtonianEuler/Sources.hpp:
```diff
@@ -0,0 +1,37 @@
+      gsl::not_null<tnsr::I<DataVector, volume_dim>*> source_momentum_density,
+      gsl::not_null<Scalar<DataVector>*> source_energy_density,
+      const db::item_type<Tags::SourceTerm<SourceTermType>>& source,
+      Args&... args) noexcept {
```
Is `Args&` intended here, or should it be `const Args&` or `Args&&`?

carmaza replied:
I think here I just chose the one that wasn't giving compilation errors. Which one should be the preferred here?

wthrowe replied:
`const Args&` should be correct if it compiles.  If that doesn't work I'll take a closer look at exactly how this is being used.

carmaza replied:
`const Args&` worked. Not sure if/why I tried that before :P

wthrowe reviewed (changes requested):
Does it make sense to "mix and match" sources and analytic data/solutions, or do the values in the metavariables have to be kept in sync?  if the latter, could the sources be specified by the analytic data class?

carmaza commented:
@wthrowe Ideally, we'd like that any parameter(s) shared both by analytic data and sources be available from a common variable. I'm open for suggestions on how to do that!

wthrowe commented:
The simplest way I can think of would be to have a method on the analytic data that returns the appropriate sources object and then use the new flexible OptionTag -> Tag mapping to construct both from the same input file section.

carmaza committed:
Rename IsentropicVortexSource VortexPerturbation

carmaza committed:
Mark IsentropicVortex constructor noexcept

carmaza committed:
Add function for computing perturbation term in IsentropicVortex

carmaza committed:
Update VortexPerturbation source to evolve modified IsentropicVortex

 - Apply function now takes coords and time as arguments
 - Arguments are used to compute vortex primitives locally
 - Source terms are computed from vortex primitives
 - Update argument and return tags

carmaza committed:
Add source term member to IsentropicVortex

carmaza committed:
Force 2D IsentropicVortex to use vanishing perturbation

carmaza committed:
Add source_term_type alias in RiemannProblem

carmaza committed:
Add NoSource type for NewtonianEuler

carmaza commented:
Regard this as a whole new PR. Commit order should be:

- Add UniformAcceleration source for NewtonianEuler
- Mark IsentropicVortex constructor noexcept
- Add function for computing perturbation term in IsentropicVortex
- Rename IsentropicVortexSource VortexPerturbation
- Update VortexPerturbation source to evolve modified IsentropicVortex
- Add source term member to IsentropicVortex
- Force 2D IsentropicVortex to use vanishing perturbation
- Add NoSource type for NewtonianEuler
- Add source_term_type alias in RiemannProblem
- Add ComputeSources for NewtonianEuler
- Add Tags for problem-specific sources for NewtonianEuler
- Add Initialization of Source Term for NewtonianEuler
- Enable source terms in NewtonianEuler

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp:
```diff
@@ -0,0 +1,91 @@
+ private:
+  template <size_t SpatialDim>
+  friend bool
+  operator==(  // NOLINT (clang-tidy: readability-redundant-declaration)
```
`NOLINT(readability-redundant-declaration)` should no lint this only for the specified argument

kidder commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp:
```diff
@@ -154,10 +154,18 @@ class IsentropicVortex : public MarkAsAnalyticSolution {
+  /// Function of `z` coordinate to compute the perturbation generating
+  /// a source term. Public so the corresponding source class can also use it.
+  template <typename DataType>
+  DataType function_of_z(const DataType& z) const noexcept;
```
`function_of_z` -> `perturbation_profile`

kidder commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp:
```diff
@@ -154,10 +154,18 @@ class IsentropicVortex : public MarkAsAnalyticSolution {
+  DataType function_of_z(const DataType& z) const noexcept;
+
+  template <typename DataType>
+  DataType dz_function_of_z(const DataType& z) const noexcept;
```
`deriv_of_perturbation_profile`

kidder commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp:
```diff
@@ -197,13 +205,12 @@ class IsentropicVortex : public MarkAsAnalyticSolution {
-    // (3D only) Extra term in the velocity along z that generates sources.
-    DataType perturbation{};
+    // (3D only) z-coordinate to compute perturbation term
+    DataType z{};
```
`z` -> `z_coord`

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp:
```diff
@@ -60,39 +82,62 @@ namespace Sources {
+
+  using argument_tags = typename detail::get_argument_tags<Dim>::type;
+
+  // Overload required for 2-D simulations, where no variable is sourced.
```
`2-D` -> `2d` and similarly for `3d`

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp:
```diff
@@ -60,39 +82,62 @@ namespace Sources {
+ private:
+  template <size_t SpatialDim>
+  friend bool
+  operator==(  // NOLINT (clang-tidy: readability-redundant-declaration)
```
`NOLINT(rereadability-redundant-declaration)`

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp:
```diff
@@ -60,39 +82,62 @@ namespace Sources {
+      const VortexPerturbation<SpatialDim>& lhs,
+      const VortexPerturbation<SpatialDim>& rhs) noexcept;
+
+  double adiabatic_index_ = std::numeric_limits<double>::signaling_NaN();
```
Instead of creating the solution every time `apply` is called, just pass it to `apply` and then you don't need these member variables

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp:
```diff
@@ -16,33 +23,112 @@
+    const gsl::not_null<tnsr::I<DataVector, 3>*> source_momentum_density,
+    const gsl::not_null<Scalar<DataVector>*> source_energy_density,
+    const tnsr::I<DataVector, 3>& x, const double time) const noexcept {
+  Solutions::IsentropicVortex<3> vortex(
```
pass the solution as an argument so it is not recomputed everytime... (see comment in hpp file)

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp:
```diff
@@ -16,33 +23,112 @@
+  Solutions::IsentropicVortex<3> vortex(
+      adiabatic_index_, vortex_center_, vortex_mean_velocity_,
+      perturbation_amplitude_, vortex_strength_);
+  const size_t size = get<0>(x).size();
```
`size` -> `number_of_grid_points`

kidder commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp:
```diff
@@ -16,33 +23,112 @@
+      get<Tags::Velocity<DataVector, 3, Frame::Inertial>>(vortex_primitives),
+      get<Tags::SpecificInternalEnergy<DataVector>>(vortex_primitives));
+
+  DataVector dz_vortex_velocity_z =
```
save an allocation by storing `dz_vortex_velocity_z` in `*source_mass_density_cons`, and then do the multiplication by `vortex_mass_density_cons` after computing the other sources

kidder reviewed (changes requested):
partial review with @nilsdeppe , got through `add source term member to IsentropicVortex`

carmaza committed:
Enable py modules to be called across NewtonianEuler test dirs

carmaza commented:
Thanks for starting reviewing this. I'll push the fixup when the review is done. In the meantime, I'm pushing an extra commit to add some `__init__.py` files to fix a test that was broken. (The absence of these files will only break runs using python 2)

wthrowe commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp:
```diff
@@ -60,39 +82,62 @@ namespace Sources {
+
+  using sourced_variables = typename detail::get_sourced_variables<Dim>::type;
+
+  using argument_tags = typename detail::get_argument_tags<Dim>::type;
```
Can these be `tmpl::conditional_t<Dim == 3, tmpl::list<stuff>, tmpl::list<>>`?

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.cpp:
```diff
@@ -38,6 +39,14 @@ IsentropicVortex<Dim>::IsentropicVortex(
                    mean_velocity, strength) {
+  if (Dim == 2) {
+    ASSERT(
+        abs(perturbation_amplitude_) < std::numeric_limits<double>::epsilon(),
```
You can enforce this when constructed from an input file by only including `PerturbAmplitude` in the `options` list in 3D.  You'll have to overload the constructor or reorder the arguments and default `perturbation_amplitude`.  (If that doesn't work for some reason this should be a `PARSE_ERROR` because it can be triggered by an input file.)

The assertion is still good to have in case someone creates the class in C++.

wthrowe commented on src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp:
```diff
@@ -0,0 +1,76 @@
+            typename ParallelComponent,
+            Requires<tmpl::list_contains_v<
+                typename db::DataBox<DbTagsList>::simple_item_tags,
+                Initialization::Tags::InitialTime>> = nullptr>
```
You don't use the initial time.  I don't think you need this Requires.

wthrowe commented on src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp:
```diff
@@ -0,0 +1,76 @@
+    using compute_tags = db::AddComputeTags<>;
+    auto source_term =
+        Parallel::get<typename Metavariables::initial_data_tag>(cache)
+            .source_term();
```
Would it make sense to have the source term in the cache along with the analytic solution?

kidder replied:
Yes, I think that makes sense

wthrowe replied:
ping

carmaza replied:
Most of the ideas devoted to this change would require some changes beyond the scope of this PR, and they would also be applicable to putting EoSs (and possibly other classes) in the global cache. These changes would certainly imply further group discussion, so, after discussing this with @kidder, I opted to leave the source in the databox for now and come back with an enhancement once we know how we want to do it with EoSs.

wthrowe replied:
Is there more needed for this than
```c++
template <typename SolutionType>
struct SourceTerm : SourceTermBase, db::SimpleTag {
  using type = typename SolutionType::source_term_type;
  using option_tags = tmpl::list<::OptionTags::AnalyticSolution<SolutionType>>;
  static type create_from_options(
      const SolutionType& analytic_solution) noexcept {
    return analytic_solution.sources();
  }
};
```
and changing some template parameters around?

nilsdeppe replied:
Yea, I think that should work. Maybe we should talk about how to implement this after the exec call? I think EoSs could be done the same or in a similar way.

carmaza replied:
That was my first attempt, but the initial data is not restricted to analytic solutions, or even to analytic data. Adding an `OptionTags` for a generic `InitialData` type might do it, but the very fact of creating the source (or the EoS) as a member of another class seems ultimately suboptimal in cases where you don't really need it to be a member (i.e., when the source doesn't depend on parameters that needed for the initial data, e.g. for numerical initial data.) I'm fine with implementing your suggestion if you think that it will make things easier to update in the future, though.

carmaza replied:
I'd be happy to discuss what's required for (and volunteer to implement) the changes for EoSs.

carmaza replied:
So after our discussion today, I implemented Will's suggestion (last fixup commit). The initialization source term commit will be dropped once I squash.

wthrowe commented on src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp:
```diff
@@ -78,6 +80,34 @@ namespace Parallel {
+template <typename Metavariables,
+          bool = not tmpl::list_contains_v<
+              tmpl::list<typename Metavariables::source_term_type>,
+              NewtonianEuler::Sources::NoSource>>
```
`cpp17::is_same_v<typename Metavariables::source_term_type, NewtonianEuler::Sources::NoSource>`, and use `tmpl::conditional_t`.  Again below.  (Same as is done with the local time stepping stuff.)

kidder commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp:
```diff
@@ -0,0 +1,164 @@
+      &ComputeSourcesProxy<SomeSourceType<Dim>>::apply, "TestFunctions",
+      {"source_mass_density_cons", "source_momentum_density",
+       "source_energy_density"},
+      {{{-1.0, 1.0}, {-1.0, 1.0}, {-1.0, 1.0}, {-1.0, 1.0}}}, used_for_size);
```
since all the bounds are the same, you can just pass a single `{-1.0, 1.0}` and change the `4` on line 144 to `1`

kidder commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp:
```diff
@@ -0,0 +1,164 @@
+  pypp::check_with_random_values<4>(
+      &ComputeSourcesProxy2<SomeOtherSourceType<Dim>>::apply, "TestFunctions",
+      {"source_momentum_density", "source_energy_density"},
+      {{{-1.0, 1.0}, {-1.0, 1.0}, {-1.0, 1.0}, {-1.0, 1.0}}}, used_for_size);
```
again a single bound

kidder commented on src/Evolution/Systems/NewtonianEuler/Tags.hpp:
```diff
@@ -87,5 +87,16 @@ struct SoundSpeedSquared : db::SimpleTag {
+  using type = SourceTermType;
+  static std::string name() noexcept { return "SourceTerm"; }
+};
+
```
add a test for  `name`

carmaza committed:
Add getter for IsentropicVortex perturbation amplitude

kidder commented:
@wthrowe please look at the fixups

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp:
```diff
@@ -127,8 +130,8 @@ class IsentropicVortex : public MarkAsAnalyticSolution {
-  using options = tmpl::list<AdiabaticIndex, Center, MeanVelocity,
-                             PerturbAmplitude, Strength>;
+  using options = tmpl::list<AdiabaticIndex, Center, MeanVelocity, Strength,
+                             PerturbAmplitude>;
```
Remove PerturbAmplitude in 2d.

carmaza replied:
Would using a `tmpl::conditional_t` suffice?

wthrowe replied:
Yes.

nilsvu commented on src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp:
```diff
@@ -157,7 +169,7 @@ struct EvolutionMetavars {
   };
 
-  using initialization_actions = tmpl::list<
+  using initialization_actions = tmpl::flatten<tmpl::list<
```
you can remove the `flatten` as of #1699

nilsdeppe replied:
ping on this.

nilsvu commented on src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp:
```diff
@@ -0,0 +1,90 @@
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& /*p*/) noexcept;  // NOLINT
+
+  using sourced_variables = tmpl::list<Tags::MomentumDensity<DataVector, Dim>,
```
If you rename `sourced_variables` -> `return_tags` and `db::wrap_tags_in<Tags::Source, ...>` then you don't need the `ComputeSources` wrapper in the other file. I don't think it's worth having that wrapper only to do the `wrap_tags_in`. The system can then directly retrieve `volume_sources = typename InitialDataType::sources`.

carmaza replied:
Considering that our generic `ComputeVolumeSources` action for evolution systems is a wrapper to the databox function `mutate_apply`, I am not sure how to make the latter work with stateful structs such as sources holding their own members.

nilsvu replied:
Ah, I see now that your `ComputeSources` wrapper also adds the `Tags::SourceTerm` to the `argument_tags` and forwards the apply call to it as a member function.
Note that `mutate_apply` also has an overload for stateful structs, i.e. the following would work:
```cpp
db::mutate_apply(get<Tags::SourceTerm<InitialDataType>>(box), make_not_null(&box))
```
We could actually specialize the stateless `db::mutate_apply<Invokable>(make_not_null(&box))` overload for the case where `Invokable` is a tag, such that it retrieves the tag and calls the stateful overload automatically. But this is out of scope for this PR, so having the `ComputeSources` wrapper seems like a good solution to me.

nilsvu replied:
I pushed an implementation of such a specialization to #1728

nilsvu commented on src/Evolution/Systems/NewtonianEuler/Tags.hpp:
```diff
@@ -87,5 +90,24 @@ struct SoundSpeedSquared : db::SimpleTag {
+/// The source term in the evolution equations
+template <typename InitialDataType>
+struct SourceTerm : SourceTermBase, db::SimpleTag {
+  using type = typename InitialDataType::source_term_type;
```
In the elliptic code I have an `elliptic::Tags::FluxesComputer`, so you could name this tag `SourcesComputer`. I don't think we have naming guidelines for these sort of tags yet, but we can try to be consistent.

nilsvu commented on src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp:
```diff
@@ -93,6 +94,7 @@ class IsentropicVortex : public MarkAsAnalyticSolution {
 
  public:
   using equation_of_state_type = EquationsOfState::PolytropicFluid<false>;
+  using source_term_type = Sources::VortexPerturbation<Dim>;
```
I think the simpler `using sources = Sources::VortexPerturbation<Dim>` is more in line with how we name these aliases e.g. in the `Metavariables`. No need to add a `_type` suffix (could also be removed from the EOS in the line above, but not in this PR).

nilsvu commented on tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp:
```diff
@@ -0,0 +1,171 @@
+      {{{-1.0, 1.0}}}, used_for_size);
+  pypp::check_with_random_values<1>(
+      &ComputeSourcesProxy2<SomeInitialData<SomeOtherSourceType<Dim>>>::apply,
+      "TestFunctions", {"source_momentum_density", "source_energy_density"},
```
If I understand correctly, this test is only for the `ComputeSources` class, which does nothing but wrap the `sourced_variables` in `Tags::Sources`. If you do that in the source class directly (see comment above about `sourced_variables` -> `return_tags`) then you don't need this test at all.

Even if you were to keep this test, I don't think you should go through a Python implementation of the sources that are _only used within this test_. Unit tests should test no more and no less than the piece of code they are designed to test. In this case, that would only be that `ComputeSources<...>` is compatible with `mutate_apply` and that it returns `Tags::Source<...>`.

nilsvu reviewed (commented):
Not a complete review, but things I noticed when reading through this PR. Looks like this PR has enough reviewers already, so don't wait for my approval for fixups or merging.

wthrowe commented:
I am happy with the fixups for my comments.

kidder commented:
I am happy with the fixups as well

carmaza commented:
Well, I'm ready to squash whenever.

nilsdeppe reviewed (commented):
I'm happy for you to squash.

carmaza committed:
Enable source terms in NewtonianEuler evolution

kidder commented:
@wthrowe @nilsdeppe this has been squashed and rebased

kidder commented:
@wthrowe if you are happy with this, please merge it

wthrowe commented on src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp:
```diff
@@ -16,33 +23,70 @@
+      get<Tags::Velocity<DataVector, 3, Frame::Inertial>>(vortex_primitives),
+      get<Tags::SpecificInternalEnergy<DataVector>>(vortex_primitives));
+
+  // We save the precomputed value of dv_z/dz in source_mass_density_cons
+  // in order to save an allocation
+  get(*source_mass_density_cons) =
+      vortex.perturbation_amplitude() *
+      vortex.deriv_of_perturbation_profile(get<2>(x));
```
A simpler way to do this would have been to reuse the allocation returned by `deriv_of_perturbation_profile`.  (Assign it to a variable and then `*=` the amplitude.)  (Just a comment.  Not requesting a change.)

wthrowe reviewed (approved):
@nilsleiffischer If there are any further changes you want, please file issues for them.

Ignoring Travis timeout.
